File: GRAMATICAS/GRAMATICA5/gramatica5.c
#include <stdio.h>
#include <string.h>

int acepta_cadena(const char *cadena) {
    int len = strlen(cadena);

    // Debe empezar con 'a' y terminar con 'b'
    if (len < 2 || cadena[0] != 'a' || cadena[len - 1] != 'b') {
        return 0;
    }

    // Caso base: "ab"
    if (len == 2 && cadena[0] == 'a' && cadena[1] == 'b') {
        return 1;
    }

    // Verificar el patrón a(ab)^n b
    // Desde el segundo carácter hasta el penúltimo deben repetirse "ab"
    for (int i = 1; i < len - 1; i += 2) {
        if (i + 1 >= len - 1) return 0; // No hay pareja completa
        if (cadena[i] != 'a' || cadena[i + 1] != 'b') {
            return 0;
        }
    }
    return 1;
}

int main() {
    char archivo[100], linea[100];
    FILE *fp;

    printf("Ingrese el nombre del archivo: ");
    scanf("%s", archivo);

    fp = fopen(archivo, "r");
    if (!fp) {
        perror("No se pudo abrir el archivo");
        return 1;
    }

    while (fgets(linea, sizeof(linea), fp)) {
        // Eliminar salto de línea
        linea[strcspn(linea, "\n")] = '\0';

        if (acepta_cadena(linea)) {
            printf("acepta\n");
        } else {
            printf("no acepta\n");
        }
    }

    fclose(fp);
    return 0;
}

File: GRAMATICAS/GRAMATICA5/gramatica5.py
def acepta_cadena(cadena):
    # Debe empezar con 'a' y terminar con 'b'
    if len(cadena) < 2 or cadena[0] != 'a' or cadena[-1] != 'b':
        return False

    # Caso base: "ab"
    if cadena == "ab":
        return True

    # Verificar el patrón a(ab)^n b
    # Desde el segundo carácter hasta el penúltimo deben repetirse "ab"
    i = 1
    while i < len(cadena) - 1:
        if i + 1 >= len(cadena) - 1:
            return False  # No hay pareja completa
        if cadena[i] != 'a' or cadena[i + 1] != 'b':
            return False
        i += 2
    return True
   
def main():
    archivo = input("Ingrese el nombre del archivo: ")

    try:
        with open(archivo, "r") as f:
            for linea in f:
                cadena = linea.strip()
                if acepta_cadena(cadena):
                    print("acepta")
                else:
                    print("no acepta")
    except FileNotFoundError:
        print("No se pudo abrir el archivo")

if __name__ == "__main__":
    main()

